spyder/utils/workers.py

```diff
@@ -0,0 +1,358 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright © Spyder Project Contributors
+#
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+# -----------------------------------------------------------------------------
+"""Worker threads for downloading files."""
+
+# Standard library imports
+from collections import deque
+import os
+import sys
+
+# Third party imports
+from qtpy.QtCore import (QByteArray, QObject, QProcess, QThread, QTimer,
+                         Signal)
+
+# Local imports
+from spyder.py3compat import PY2, to_text_string
+
+
+WIN = os.name == 'nt'
+
+
+def handle_qbytearray(obj, encoding):
+    """Qt/Python2/3 compatibility helper."""
+    if isinstance(obj, QByteArray):
+        obj = obj.data()
+
+    return to_text_string(obj, encoding=encoding)
+
+
+class PythonWorker(QObject):
+    """
+    Generic python worker for running python code on threads.
+
+    For running processes (via QProcess) use the ProcessWorker.
+    """
+    sig_started = Signal(object)
+    sig_finished = Signal(object, object, object)  # worker, stdout, stderr
+
+    def __init__(self, func, args, kwargs):
+        """Download Worker based on requests."""
+        super(PythonWorker, self).__init__()
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self._is_finished = False
+        self._started = False
+
+    def is_finished(self):
+        """Return True if worker status is finished otherwise return False."""
+        return self._is_finished
+
+    def start(self):
+        """Start the worker (emits sig_started signal with worker as arg)."""
+        if not self._started:
+            self.sig_started.emit(self)
+            self._started = True
+
+    def terminate(self):
+        """Mark the worker as finished."""
+        self._is_finished = True
+
+    def _start(self):
+        """Start process worker for given method args and kwargs."""
+        error = None
+        output = None
+
+        try:
+            output = self.func(*self.args, **self.kwargs)
+        except Exception as err:
+            error = err
+
+        if not self._is_finished:
+            self.sig_finished.emit(self, output, error)
+        self._is_finished = True
+
+        
+class ProcessWorker(QObject):
```

Comment:
@jitseniesen this is the idea for the process worker, maybe you could take a look and see if it fits your needs.